### PR TITLE
Display - symbol when tracked in-game leaderboard score is below 50

### LINKED ISSUE #21176: Display `-` symbol when tracked in-game leaderboard score is below 50
### Type

Cosmetic

### Bug description

When there are more than 50 scores already set, gameplay leaderboard shows at the beggining that the player is 51st

image

This would imply that with the score that player has currently, they would end up with 51st place on the map's leaderboard, which is not always true.

What I would suggest is to change the current `#51` position to `?` or something, untill player beats the 50th score.

### Screenshots or videos

_No response_

### Version

2022.1101.0-lazer

### Logs

[database.log](https://github.com/ppy/osu/files/9960974/database.log)
[input.log](https://github.com/ppy/osu/files/9960975/input.log)
[network.log](https://github.com/ppy/osu/files/9960976/network.log)
[performance.log](https://github.com/ppy/osu/files/9960977/performance.log)
[runtime.log](https://github.com/ppy/osu/files/9960978/runtime.log)

### PR DESCRIPTION
**Pull request by Piggey**
This PR aims to fix https://github.com/ppy/osu/issues/21176

When there already are 50 submitted scores, player's position becomes `-` instead of `#51` , until player beats 50th score

image

This happens only at 50 submitted scores, meaning when there are for example 51 local scores, gameplay leaderboard will display `#52` as players position

There's one case however when there are exactly 50 local scores, leaderboard will display `-` even though `#51` would be technically correct.

I believe that one case where this solution doesn't work is still much better than the solution used before, when only one case was correct.

### CODE REVIEW

**Comment by loma-baldsson** [comment]
This seems very hacky. What if, in the future, more than 50 scores are returned from the leaderboards? What if there are exactly 50 scores on the leaderboard, thus making your current score # 51? What if someone just has 50 local scores? I don't think your code accounts for any of these scenarios

**Comment by smoogipoo** [comment]
I think the future path forward is to show the true position of this score during gameplay, which would make this PR unnecessary. Is that right @peppy ?

**Comment by peppy** [comment]
> I think the future path forward is to show the true position of this score during gameplay, which would make this PR unnecessary. Is that right @peppy ?

As per discussion in #21176 (and https://github.com/ppy/osu/discussions/20538 by extension), that's correct yeah. I did say that this would be nice as a temporary thing, but the implementation here is more hacky than we can even warrant for a temporary solution.

File: osu.Game/Screens/Play/HUD/GameplayLeaderboard.cs [hunk 1]
```diff
@@ -174,6 +174,10 @@ private void sort()
                 orderedByScore[i].ScorePosition = i + 1;
             }
 
+            // change displayed potision to '-' when there are 50 already submitted scores and tracked score is last
+            if (TrackedScore?.ScorePosition == Flow.Count && Flow.Count == 51)
+                TrackedScore.ScorePosition = null;
```
Comment (peppy) [inline osu.Game/Screens/Play/HUD/GameplayLeaderboard.cs:179]: This should be based on when the score is the last in the flow, not specifically for a count of "51".

To make this work, you'll also need to differentiate a "solo"/"playlist" leaderboard from a multiplayer one, to handle the case where a multiplayer room always shows all users.

**Review by peppy — CHANGES_REQUESTED** [review]

File: osu.Game/Screens/Play/HUD/GameplayLeaderboard.cs (hunk 1 above)
Comment (bdach) [reply osu.Game/Screens/Play/HUD/GameplayLeaderboard.cs:179]: The multiplayer part of this comment is valid, but I'm wondering whether checking if the item is last in the flow is 100% correct? I'm thinking about the case where there are 49 or less submitted score for a given beatmap, at which point showing 50 for the play currently in progress would technically not be wrong.

Comment (peppy) [reply osu.Game/Screens/Play/HUD/GameplayLeaderboard.cs:179]: Valid point. Checking `>=50` could work if that's referencing a constant that's also sent to the web request somehow.

**Comment by Piggey** [comment]
So I've moved the logic to the `SoloGameplayLeaderboard` class since it doesn't really affect other leaderboards.

It now displays values as before for local leaderboards, but displays " - " for every mode apart from local.
I'm using `OsuSetting.BeatmapDetailTab` to differentiate "local" leaderboard scores from "global" ones, which could be finicky since the "details" tab can display both local and global scores.

As for the maximum number of global scores I'm completely lost. I feel like making a web request each time a leaderboard is created is pointless, since the actual number doesn't change that often. Could this value be stored somewhere?

**Comment by peppy** [comment]
I've made the remaining changes required to get this in. Please take a look through my commits so you can see what was required.

**Review by peppy — DISMISSED** [review]

**Review by peppy — APPROVED** [review]

**Pull request merged**